provider/core/src/varule_traits.rs

```diff
@@ -0,0 +1,121 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use zerovec::ule::VarULE;
+
+#[cfg(feature = "alloc")]
+use zerovec::{maps::ZeroMapKV, ZeroMap2d};
+
+/// A trait that associates a [`VarULE`] type with a data struct.
+///
+/// Some data structs can be represented compactly as a single [`VarULE`],
+/// such as `str` or a packed pattern. This trait allows for data providers
+/// to use storage optimizations for such types.
+///
+/// ❗ Not all data structs benefit from this optimization. It is best when the
+/// data struct is multiplied across a large number of data marker attributes.
+///
+/// Both [`MaybeAsVarULE`] and [`MaybeEncodeAsVarULE`] should be implemented
+/// on all data structs. The [`data_struct!`] macro provides an impl.
+pub trait MaybeAsVarULE {
+    /// The [`VarULE`] type for this data struct, or `[()]`
```

Comment:
ok, let's have this conversation here then:

why use `[()]`?

Reply:
Right. I'm fine with a zero-sized VarULE type, and I'm fine with `[()]` or `[DummyZST]` or something, the former concerns me a _little bit_ more but I already went through our code and I think it's not risky.


Uninhabited types are a whole other ballgame and we should not be introducing those.